README.md

**27th July 2023**

* Adjust mesowx.py according to Issue #8
   https://github.com/glennmckechnie/weewx-mesowx/issues/8
   It seems that Thread.isAlive() was deprecated in Python 3.7 in favor of is_alive()
   (which has been around since Python v2.6), and removed in Python v3.9.
* While this is an untested fix (by me - I'm still running python 3.7) I note that
  this fix exists in the main weewx code base, and it obviously works for MaffooClock.
  And Thank you MaffooClock for the watchful eyes and heads up.


**17th Sept 2022**

* Code tidy up. 

* Add try for missing server "ConnectionError" fallout. No longer fatal, it should keep running now.
Pending release - updated code is in repo master.

* The original code at https://bitbucket.org/lirpa/mesowx has gone. So this is it!

**25th June 2020**
### Languages

Released as [version 0.6.4](https://github.com/glennmckechnie/weewx-mesowx/releases/tag/v0.6.4)

* text labels are now configurable to accept alternate names or languages.
    mesowx.py, index.html.tmpl, skin.conf and Config.js.tmpl have been tweaked to
    1.) retain a set of english defaults
    2.) read values from the [Language] sections in skin.conf

* No change to skins/mesowx/XXX.inc files, they will still need to be hand edited. They are
    intended to be user configurable anyway.


### Better Bling!

Released as [version 0.6.3](https://github.com/glennmckechnie/weewx-mesowx/releases/tag/v0.6.3)

* split unit configuration into database (auto via weewx.conf) **and** display
    (manual via skin.conf)
* issues ref #2 : change displayed unit inHg for barometer to 'inHg' instead of 'in'
* flag (stats - min/max) colors are now the same as the line colors. They should also be
    visible when near the boundaries (clip : false)
* add optional inTemp and inHumidity sections to mesowx console. We have the room, if we
    have the fields...
* re-add rain to charts. Make dayRain a configurable option from skin.conf (untested, don't
    have one, feed back welcomed!)
* Ratchet up the user configuration in skin.con - 2 default pallets and the previous user
    configurable one
* add chart visibility options to skin.conf. All but inTemp are 'on' to start with.
    outTemp is always on (otherwise the chart generation errors and we get nowhere)
* add modules/exporting.js to add print menu (top right hamburger style)
* there's still an issue #4 with the Humidity flags staying visible
* tweak dayRain config option (untested - no Davis here)

----

Updated the d3, jquery, and highstock libraries.

Adjust css, _index.html_ and _Config.js_ to improve screen usage and allow dynamic rescaling.

Add colors to _sink.conf_ match them to _Config.js_ and update highcharts options to match 8.1.1


See the CHANGELOG.md for a brief summary of changes.

----

**Jun 2020:  Weewx and wee_extension -- DONE!**

The move to use *wee_extension* for the installation process is done. This, along with various updates and easier configuration gives a working default local installation and a far easier process to set up a remote installation (move a directory, rename one file). Testing and feedback is always welcome...

More could be done with this skin and with support and contributions from the community that could still happen. Fork away, contribute back, raise issues. It's not dead yet (it's just sleeping squire.)
# [WeeWX-MesoWx](https://github.com/glennmckechnie/weewx-mesowx)

----

This repo is a fork of MesoWx as originally implemented at [Mesowx](https://bitbucket.org/lirpa/mesowx)

## What is MesoWx

MesoWx is a real-time HTML front-end for visualizing personal weather station data. It provides a real-time graph and console display, and dynamic graphs of your weather station history allowing you to explore the details of any recorded time period in your data.

MesoWx displays data from a database and does not itself interface with any weather station hardware directly, however, being built upon Meso it supports an HTTP API for remotely adding data, which allows integration with existing weather station software. MesoWx integrates well with Weewx and should support any weather station that it supports.

![v0.6.4-1month-archive](lirpa/images/v0.6.4-1month-archive.png)
![Samsung-landscape-v064.png](lirpa/images/Samsung-landscape-v064.png)

**June 2020:** Time has passed and there are many more excellent and capable skins available for weewx now. They have evolved with the community and are not only well integrated with weewx but feature packed. This one lacks the bells and whistles but is still quite capable at conveying the collected information, thanks to highcharts, decent hardware :-) and access to the whole database.


## History

This repo started with the 0.4.0 version which was the latest available on the 1st February 2018. It has been copied here to enable integration with Luc's raw.py and sync.py scripts
No license was available at the time of upload but Peter Finley (MesoWx author) has indicated that it is [free to use](https://groups.google.com/d/msg/meso-user/ebs6sOhNqsg/iNeqnVarEgAJ)

The lirpa directory now contains the updated scripts as written by Luc Heijst and that were updated/rewritten to keep pace with either the changes in weewx versions, or issues that arose and were reported by weewx users.
Some of those scripts are available in various posts on the weewx-users group. The versions that Luc passed on have been uploaded here as individual commits so viewing the commit history of this repo will show some of the history of the mesowx scripts, after the lirpa repo stopped  development.

Those scripts and associated contents have been tweaked to enable the weewx installer to control the installation process. The use of wee_extension vastly improves the installation process and mesowx configuration and should go someway towards keeping MesoWx alive and current.

Previously it took a lot of effort to modify Mesowx to personal taste. The default local (Raw) installation will just run once installed and the RemoteSync version is in a state where a cut, paste and a file rename will get it into a working state (the database setup does involve a few more steps. Ah well!) Configuration is predominately via _weewx.conf_ and _skin.conf_

## Limitations

 1. A modern web browser is required for the out of the box front-end (Chrome, Firefox, Safari, IE10+)
 2. Extremely large archive databases likely won't perform very well at the moment (my archive database contains 350,000+ records and has performed adequately), especially on a low power server.
 3. To record the raw (loop) packets it uses mysql, these days MQTT has filled that niche and is far better suited to it. This skin does not use MQTT.

## Warnings

This software may be dated, but it is still functional. It fills a niche for me, and possibly for others. It was also a long term objective on my part to see if it could be managed by *wee_extension*, and that's one itch that has now been scratched!

The old caveat of there being **NO** update path still applies. However the new script - _mesowx.py_ - hasn't really changed that much, only the way it's called. If you have a working install then don't blow that away, the script should still work with it, you'll just need to call different sections of it if you decide to update to this weewx4/python3 compatible version.
So, Backup your existing version, don't overwrite it. Install this one in parallel, in its new location, and then copy your changes over. If it's index.html changes then a drop in could work. If it's the config files then weewx.conf will need tweaking. If it's the script then you'll need some python knowledge to sort out the redirection.

If that all seems too hard then contact me via github and we'll see what we can work out.

----

# Weewx and wee_extension -- DONE!

**Jun 2020**
The move to weewx via *wee_extension* is done, it needs some beta testing...

*wee_extension* will install and configure this skin as much as it can. More could be done with the integration and with support and contributions from the community that could still happen.

The 3 original mesowx scripts have been combined into one script which are all installed as a SLE - a skin named Mesowx, a script named _mesowx.py_ and a database named mesowx.

Because it has control over the local setup ... **[Mesowx][[Raw]]** ... you will find that version fully configured and ready to view in your browser at http://localhost/weewx/mesowx/

The skin is populated with values sourced from _weewx.conf_ and from _install.py_. It applies them to both the Raw and RemoteSync versions.

It also generates the two unique security keys that are required for the RemoteSync version.

# Install, then configure

I've been more rigourous with packaging up the point releases as something meaningful and trackable (especially if I break something). That's made that page fairly busy but it offers a good way of tracking the changes, especially for the switch to v0.6.X where the change to d3, highcharts and jquery library was made.

    wget -O weewx-mesowx.zip https://github.com/glennmckechnie/weewx-mesowx/archive/master.zip
    wee_extension --install weewx-mesowx.zip

## Configuration values in weewx.conf and skin.conf

The supplied values, except for the **remote_server_url** and the **polling_interval** can all be left as they are.

Your values for the *raw_security_key* and *archive_security_key* will be different. Those are auto generated by the installer and will be unique for every installation. You can change them if you like, and then wait for your values to propagate through to the webserver files - all before you transfer the files etc. to the remote server.

The remote server URL obviously needs changing.

The polling interval sets the refresh time in seconds (converted to milliseconds) 60 is the default. I use 20 for the 1-wire setup. It's up to you and your station.

**skip_loop** sets the time allowed between loop packets. If you get them every 2 seconds and you want to change that to every 20 seconds, then the 2 becomes 20.
If you already get them every 20 seconds, then the default value of 2 seconds will have no affect. Leave it or increase it if 20 seconds is still too frequent. It's your choice.

The **data_limit** is the number of hours of records to keep in the raw database, anything older than that are deleted. It is also used in the 24-hour range. If you opt for 48 then it should display 48 hours of raw records (rather than the default of 24-Hours)


    # Options for extension 'mesowx'

    [Mesowx]
       loop_polling_interval = 60
       [[RemoteSync]]
            raw_entity_id = weewx_raw
            remote_server_url = http://remote_server_address/weewx/mesowx/meso/
            raw_security_key = gQVAtmzFP6XkwlCdiUi
            archive_entity_id = weewx_archive
            archive_security_key = w29VxwOOzJYQxnKa3h
        [[Raw]]
            data_binding = mesowx_binding
            data_limit = 48
            skip_loop = 2

ChartColors is a dict of values in the _Mesowx/skin.conf_ file which are used by the SLE (CheetahGenerator ) to populate the colors in _skins/Mesowx/js/Config.js.tmpl_, which then becomes _mesowx/js/Config.js_
You can now easily get rid of my values and choose your own!

    [ChartColors]
        # These are the color codes used in Config.js to define the chart lines and
        # the chart names.
        # To change these colors, and have them take affect, you will need to have
        # the skin set to enable in weewx.conf, ie:- under [Mesowx] enable = True
        #
        # One of many ... https://www.rapidtables.com/web/color/Web_Safe.html

        out_temp = '#0033FF'
        bar_ometer = '#660000'
        wind_speed = '#CC3300'
        wind_dir = '#999900'
        r_ain = '#80699B'
        rain_rate = '#89A54E'
        out_humidity = '#3D96AE'
        in_temp = '#CC00CC'
        dew_point = '#8BBC21'
        wind_chill = '#0d233A'
        heat_index = '#FF9900'
        wind_gust = '#336666'
        wind_gustdir = '#B5CA92'
        day_rain = '#0D233A'
        in_humidity = '#F28F43'


## Local Installation.

The local (Raw) mesowx database will be automatically generated and consists of just the raw (loop) table. The archive (REC) values will be sourced from your existing mysql weewx database.

Once all the mesowx files are generated (at last count there were 51 of them) and transferred to your local web server the **[StdReport][[Mesowx]]** skin can be disabled by changing the _enable = True_ stanza to...

    enable = False

#### Why False?

A working installation does not require that skin to be regenerated each report interval.

If the defaults work for you then it can be disabled after the first 1 or 2 cycles.

However, if you want to tinker with colors (available via _skin.conf_), change any of the **[Mesowx]** entries in weewx.conf, or experiment with the server files (remember the _XXXX*.tmpl_ files are the source files) then you'll need to renable the skin to allow those config files to be recreated, or your new files to be the source files.
If you make changes to the webserver files with the skin disabled (_enable = False_) then those changes run the risk of being overwritten if you re-enable the skin and forget what you've done (time passes for us all and memories fade!).

## Remote Installation.

The local (Raw) installation is setup to run once weewx is restarted.

The RemoteSync installation is not. To get weewx to run this section of the script you will need to manually edit the weewx.conf file by appending **user.mesowx.SyncService** to the end of restful_services, located under the **[Engine][[Services]]** section

    [...]

    [Engine]
        [[Services]]
              [...]
              restful_services = [...], user.mesowx.SyncService
              [...]

If and when you uninstall MesoWX, that entry will need to be removed.

Likewise, if you don't want the local (Raw) version of Mesowx running then remove...

    , user.mesowx.RawService

from the end of the archive_services = [...] line.

The RemoteSync version will also need action on your part as all the files that are generated and transferred to the local web server need to be transferred to your remote web server. This is the exact same directory and structure as the local, Raw version uses.  (ie:- That _www/html/weewx/mesowx_ directory is fully portable.)

Copy that whole mesowx directory to your remote webserver and then; once all the files and directories are moved to that remote machine; then rename _config-RemoteSync.json_ to _config.json_ to allow that remote installation to work correctly.

ie:- the file...

       mesowx/meso/include/config-RemoteSync.json

needs to be renamed as...

       mesowx/meso/include/config.json

With that done you should have a working webserver installation; once the database is created and populated with data then you will know for certain.

### Databases and Passwords

The "host", "user", "password", and "database" configuration values preset in mesowx/meso/include/config.json are provided to give you a working installation on a machine you have total control over. 

In this first case, where it could be a home network, or a server in your DMZ; the only setting that is system specific is "host". That needs to match what your mysql installation expects and that's most likely to be "localhost". The "user", "password", and "database" values are entirely up to you. Change them if you want to, or leave them as they are.

On the otherhand, If you are using a hosting provider that provides their own generated values, ones that are specific to your account, then you obviously need to use those values.
Edit the config.json file you renamed above and upload that, overwriting the previous version.

On the subject of hosting providers. You get what you pay for and at least one of the 'free' hosting providers won't allow a service such as mesowx to run (on their free accounts). See [issue #6](https://github.com/glennmckechnie/weewx-mesowx/issues/6) at the github site for the full details but basically they intercept the call to data.php and inject there own code (a cookie generator) which mesowx doesn't expect, and can't deal with (as the providers intended).

Also, keep in mind the traffic usage with the Free providers. You're uploading your archive records every archive interval, and your LOOP packets as often as they are generated. This has the potential to add up very quickly and if you are rate limited then you may hit the wall and the only sign will be a gap/s in data.

## Enhance or Break??

Okay, You now have a working setup, but it makes a few assumptions about fields and units. These can all be modified but you might end up with some breakage.
As of version 0.5.1 the installer will adjust the units it can according to the configuration of the target_units setting defined in weewx.conf ie: US, METRICWX or METRIC. You may want to refine them further and it is here that it is done (after the skin is disabled, otherwise your edits will be overwritten.)

As configured, it will only use the database fields as hardcoded within the file _config.json_

They are listed under "columns", in two places. Here's the one at line 58 for the archive table...

            "columns" : {
                "dateTime" :    {"type" : "number", "unit" : "s"},
                "interval" :    {},
                "barometer" :   {"unit" : "$press"},
                "inTemp" :      {"unit" : "$degr"},
                "outTemp" :     {"unit" : "$degr"},
                "inHumidity" :  {"unit" : "perc"},
                "outHumidity" : {"unit" : "perc"},
                "windSpeed" :   {"unit" : "$speed"},
                "windDir" :     {"unit" : "deg"},
                "windGust" :    {"unit" : "$speed"},
                "windGustDir" : {"unit" : "deg"},
                "rainRate" :    {"unit" : "$rainR"},
                "rain" :        {"unit" : "$meas"},
                "dewpoint" :    {"unit" : "$degr"},
                "windchill" :   {"unit" : "$degr"},
                "heatindex" :   {"unit" : "$degr"}
            },

then again at line 107 for the raw (loop) values.  The units will default to the value detected in _weewx.conf_

If you are going to add fields that will be fine, you don't have to use them. However those edits need to be done before you create the remote database or allow it to be populated. It can't create what it doesn't know about.

The other location that will need to be modified is under _mesowx/js/Config.js_ and these units and there formatting will also be modified according to the weewx defaults in _weewx.conf_

    Config.fieldDefaults = {
      'dateTime':     new meso.FieldDef('dateTime',   meso.Unit.ms,       0,    meso.Agg.avg),
      'outTemp':      new meso.FieldDef('outTemp',    mesowx.Unit.$degr,  1,    meso.Agg.avg, 'Out Temp'),
      'dewpoint':     new meso.FieldDef('dewpoint',   mesowx.Unit.$degr,  1,    meso.Agg.avg, 'Dewpoint'),
      'rain':         new meso.FieldDef('rain',       mesowx.Unit.$meas,  $m_f, meso.Agg.sum, 'Rain'),
      'rainRate':     new meso.FieldDef('rainRate',   mesowx.Unit.$rainR, $m_f, meso.Agg.max, 'Rain Rate'),
      'dayRain':      new meso.FieldDef('dayRain',    mesowx.Unit.$meas,  $m_f, meso.Agg.max, 'Rain Today'),
      'windSpeed':    new meso.FieldDef('windSpeed',  mesowx.Unit.$speed, 0,    meso.Agg.avg, 'Wind Speed'),
      'windDir':      new meso.FieldDef('windDir',    mesowx.Unit.deg,    0,    meso.Agg.avg, 'Wind Direction'),
      'windGust':     new meso.FieldDef('windGust',   mesowx.Unit.$speed, 0,    meso.Agg.avg, 'Wind Gust'),
      'windGustDir':  new meso.FieldDef('windGustDir',mesowx.Unit.deg,    0,    meso.Agg.avg, 'Wind Gust Direction'),
      'outHumidity':  new meso.FieldDef('outHumidity',mesowx.Unit.perc,   1,    meso.Agg.avg, 'Out Humidity'),
      'barometer':    new meso.FieldDef('barometer',  mesowx.Unit.$press, $p_f, meso.Agg.avg, 'Barometric Pressure'),
      'windchill':    new meso.FieldDef('windchill',  mesowx.Unit.$degr,  1,    meso.Agg.avg, 'Wind Chill'),
      'heatindex':    new meso.FieldDef('heatindex',  mesowx.Unit.$degr,  1,    meso.Agg.avg, 'Heat Index'),
      'inTemp':       new meso.FieldDef('inTemp',     mesowx.Unit.$degr,  1,    meso.Agg.avg, 'In Temp'),
      'inHumidity':   new meso.FieldDef('inHumidity', mesowx.Unit.perc,   1,    meso.Agg.avg, 'In Humidity')
     };

If you change any of these defaults, by editing these files directly; then be very careful as typos can be silent code breakers.

Take a backup (as you always do ?) and use a lot of care.

### I tweaked some values and "It doesn't work"

See the Troubleshooting section below. :-)

## Testing the default installation
                 (ie: excluding the above modifications).

A quick test of the configuration is to point the browser to the remote site using...

http://remote_server_address/weewx/mesowx/meso/data.php?entity_id=weewx_archive&data=dateTime&order=desc&limit=1

and a date should be returned if the database has something...

       [[1592045160]]

A database with nothing in it will return...

       []

A database that doesn't exist will throw a Fatal error...

       Fatal error: Uncaught PDOException: SQLSTATE[HY000] [1049] Unknown database
       [...]
        /var/www/html/weewx/mesowx/meso/include/PDOConnectionFactory.class.php on
       line 31

To create the remote database and give the appropriate permissions that match the wee_extension installation values, then...

       mysql -uroot -p
       create database mesowx;
       GRANT select, update, create, delete, insert ON mesowx.* TO mesowx@'localhost' IDENTIFIED BY 'weewx';
       quit;

The remote mysql database - mesowx - contains 2 tables; archive and raw (weewx via the _mesowx.py_ script, will create those).

A large database will take a lot to create via the backfill method.

For this reason the archive section should be copied from your existing weewx database via a mysqldump command, or other means.

If that doesn't worry you, or if it's a machine that you can see from your chair (ie: it's local and an ultra low ping away) then mesowx.py can be used to backfill an empty database. It can take a long time (just how long is a piece of string?) to complete when you start from scratch.

If it's too large it may even refuse, or your machine may be brought to it's knees while the database is being queried (admittedly if you survive that step it does get better).

While it's doing that long backfill, weewx will be stalled, that's no records generated, no graphs, nothing, nada, zilch. The only activity may be in your logs - if debug is turned on.

Once the database is populated (or starting to be) then pointing your browser to the remote machines webserver _mesowx/index.html_ file should result in a working mesowx website.

# To repeat:
           Use wee_extension to install MesoWX.
           Enable user.mesowx.SyncService by adding it to restful_services
           Restart weewx
           Wait for a report cycle, or two, to generate the <web server>/mesowx
               directory
           Copy that webserver mesowx directory and its sub directories and files
               from the local webserver and place them into the remote servers
               path - keep the same directory paths as the local installation,
               unless you know what you are doing.
           That will have created the remote website. Make sure the permissions
               are correct.
           Create the remote database, with permissions.
           Populate the remote database with the archive table from weewxs
               database, (that's all it requires from that database. The raw table
               will be created when he first loop value hits it.)
           Check the weewx.conf values are correct - pay attention to the site URL
               (Review the above hint)
           Turn off the skin generation, with a working setup it is no longer
              required. [StdReport] [[Mesowx]] enable = False
           Restart weewx with debug turned on (debug = 1)

CHANGELOG.md
# MesoWx Change Log #

## 0.6.5 (2023-07-27) ##

* Playing Catchup..
* Untested fix (by me - I'm still running python 3.7)
   Adjust mesowx.py according to Issue #8
   https://github.com/glennmckechnie/weewx-mesowx/issues/8
   It seems that Thread.isAlive() was deprecated in Python 3.7 in favor of is_alive()
   (which has been around since Python v2.6), and removed in Python v3.9.
* Also an older fix which I apparently uploaded, but didn't note here...
   rejig to stop backfill but retry on data

## 0.6.4 (2020-06-27) ##

* text labels are now configurable to accept alternate names or languages.
    mesowx.py, index.html.tmpl, skin.conf and Config.js.tmpl have been tweaked to
    1.) retain a set of english defaults
    2.) read values from the [Language] sections in skin.conf
    skins/mesowx/*.inc files will still need to be hand edited as they are intended
    to be user configurable anyway


## 0.6.3 (2020-06-23) ##

* split unit configuration into database (auto via weewx.conf) **and** display
    (manual via skin.conf)
* issues ref #2 : change displayed unit inHg for barometer to 'inHg' instead of 'in'
* flag (stats - min/max) colors are now the same as the line colors. They should also be
    visible when near the boundaries (clip : false)
* add optional inTemp and inHumidity sections to mesowx console. We have the room, if we
    have the fields...
* re-add rain to charts. Make dayRain a configurable option from skin.conf (untested, don't
    have one, feed back welcomed!)
* Ratchet up the user configuration in skin.con - 2 default pallets and the previous user
    configurable one
* add chart visibility options to skin.conf. All but inTemp are 'on' to start with.
    outTemp is always on (otherwise the chart generation errors and we get nowhere)
* add modules/exporting.js to add print menu (top right hamburger style)
* there's still an issue with the Humidity flags when the lines are hidden ??
* tweak dayRain config option (untested)

## 0.6.2 (2020-06-18) ##

* Add colors to skin.conf, match them to Config.js
* Update old highcharts options to match 8.1.1, or close enough too.

## 0.6.1 (2020-06-17) ##

* modify css, index.html, and Config.js files to allow for dynamic resizing and better use of
    screen space.
    Scales well under Opera, Firefox, Chrome and a Samsung phone. Trust that it works for others
    as well!

## 0.6.0 (2020-06-16) ##

* update to d3-v3.5.17.min.js (last of the v3 line) : highstock-v8.1.1.js : jquery-3.5.1.min.js
    libraries.
  This gives better reloads for the local version, it was stalling on the archive page for me
  There is a slight change in appearance (the navigator bar for one)
* add marker symbols etc supplied by user "laki1" back in 2017 ! ...
    https://groups.google.com/d/msg/weewx-user/eAUsTqR8yYQ/qCtf7IGaBwAJ
* for readability: split labels between left and right sides of charts

## 0.5.3 (2020-06-16) ##

* Add skip_loop as an option to skip loop (raw) records if they come too quickly for your liking
* Allow the 24-Hour display to span a longer range when specified by data_limit, assuming that
    becomes greater than the default value of 24 hour
* Set the archive graphs to load 1 months as the default. Faster initial loads

## 0.5.2 (2020-06-15) ##

* Add loop_polling_interval to [Mesowx] section. Default is 60 seconds (converts to
    60000 milliseconds for insertion into Config.js) This sets the refresh rate of the
    dashboard (right hand panel on index.html)

    From Config.js
    // the polling interval in milliseconds for the raw real-time data provider
    // this controls how frequently to fetch new data, it should be·
    // set to the frequency that your station generates "LOOP" packets
    Config.realTimePollingInterval = 60000;
* Tidy up / finish unit and format auto configurations from v0.5.1

## 0.5.1 (2020-06-15) ##

* Auto set units according to the configuration found in weewx.conf - US, METRICWX or METRIC

## 0.5.0 (2020-06-14) ##

### Simplify / Automate the install
* Combine the 3 scripts into the one - mesowx.py
* Works with weewx4 and python 3, also with python2.7
* Use the wee_extension installer to perform a (mostly) hands off install of the local setup,
    and minimize the interaction required for the remote install.
* Adjust the script, add tmpl's, use randompassword generator for security key insertion.
* Remote install works with archive and raw, Local works with raw database and weewx database,
    RetainLoopValues remains untested.

## 0.4.0 (2014-05-09) ##

### Features
* Meso version bump to 0.4.0
* Charts
    * Charts will now display minimum/maximum flags for the given time range
    * The aggregation periods are now deterministic and can be configured
    * Tooltip will now display the time range of the point if the data has been aggregated
    * Highstock version bump to 1.3.10
* No longer displaying the dayRain field by default in the console or charts since the majority
  of stations don't include this field
* Adding hPa and kPa units
* Weewx Plugins
    * retain.py
        * Can now be configured to exclude fields from being retained
    * sync.py 
        * Will now take advantage of HTTP keep-alive if your web server supports it
        * Will now keep trying if an HTTP request returns a 500 status instead of failing

### Bug Fixes
* Increasing the height of the chart to give room in case the legend wraps to two rows
* Attempting to prevent the wind speed axis scale from being needlessly large by setting it's
  maxPadding to 0

### Breaking Changes
* PHP 5.3 is no longer supported, 5.4+ is now required
* No longer support SQLite, use MySQL instead
* The sync.py weewx plugin is now dependent upon the urllib3 python library
* The chart 'numGroups' configuration parameter is no longer supported


## 0.3.4 (2014-02-19) ##

### Bug Fixes
* Fixing issue with retain.py plugin that wasn't making a copy of the retained packet
* Fixing the labels of the rain fields in Config-example.js



install.py
# Installer for weewx-mesowx (MesoWX)
# Distributed under the terms of the GNU Public License (GPLv3)
# Copyright 2018 - 2020 by Glenn McKechnie
#
# https://github.com/glennmckechnie/weewx-mesowx

from setup import ExtensionInstaller
import random
import string


def loader():
    return MesowxInstaller()


class MesowxInstaller(ExtensionInstaller):
    def __init__(self):
        # generate security keys for access to RemoteSync database
        def random_password():
            chars = string.ascii_uppercase + string.ascii_lowercase + string.digits
            size = 12
            return ''.join(random.choice(chars) for x in range(size, 30))
        arch_key = random_password()
        raw_key = random_password()

        super(MesowxInstaller, self).__init__(
            version='0.6.5',
            name='mesowx',
            description='Weather station console web front-end with'
                        ' real-time updating and dynamic charting.',
            author='Peter Finlay, modified for weewx installation'
                   'by Glenn McKechnie',
            author_email="<glenn.mckechnie@gmail.com>",
            archive_services='user.mesowx.RawService',
            config={
                'StdReport': {
                    'Mesowx': {
                        'HTML_ROOT': 'mesowx',
                        'skin': 'Mesowx',
                        'enable': 'True'}},
                'Mesowx': {
                    'loop_polling_interval': '60',
                    'Raw': {
                        'data_binding': 'mesowx_binding',
                        'data_limit': '48',
                        'skip_loop': '2'},
                    'RemoteSync': {
                        'archive_entity_id': 'weewx_archive',
                        'archive_security_key': arch_key,
                        'raw_entity_id': 'weewx_raw',
                        'raw_security_key': raw_key,
                        'remote_server_url': 'http://192.168.0.100/weewx/mesowx/meso/'}},
                'DataBindings': {
                    'mesowx_binding': {
                        'database': 'mesowx_mysql',
                        'table_name': 'raw',
                        'manager': 'weewx.manager.Manager',
                        'schema': 'user.mesowx.schema'}},
                'Databases': {
                    'mesowx_mysql': {
                        'database_type': 'MySQL',
                        'database_name': 'mesowx'}}},
            files=[('bin/user', ['bin/user/mesowx.py']),
                   ('skins/Mesowx', [
                    'skins/Mesowx/skin.conf',
                    'skins/Mesowx/js/ArchiveChart.js',
                    'skins/Mesowx/js/Compass.js',
                    'skins/Mesowx/js/Config-example.js',
                    'skins/Mesowx/js/Config.js.tmpl',
                    'skins/Mesowx/js/MesoWxApp.js',
                    'skins/Mesowx/js/MesoWxConsole.js',
                    'skins/Mesowx/js/mesowx.js',
                    'skins/Mesowx/js/MesoWxWindCompass.js',
                    'skins/Mesowx/js/RawChart.js.tmpl',
                    'skins/Mesowx/js/RealTimeChart.js',
                    'skins/Mesowx/js/WindCompass.js',
                    'skins/Mesowx/js/lib/d3-v3.5.17.min.js',
                    'skins/Mesowx/js/lib/highstock-v8.1.1.js',
                    'skins/Mesowx/js/lib/jquery-3.5.1.min.js',
                    'skins/Mesowx/js/lib/modules/exporting.js',
                    'skins/Mesowx/meso/data.php',
                    'skins/Mesowx/meso/stats.php',
                    'skins/Mesowx/meso/updateData.php',
                    'skins/Mesowx/meso/include/.htaccess',
                    'skins/Mesowx/meso/include/Agg.class.php',
                    'skins/Mesowx/meso/include/AggregateParameterParser.class.php',
                    'skins/Mesowx/meso/include/AggregateQuery.class.php',
                    'skins/Mesowx/meso/include/AggregateQuerySpec.class.php',
                    'skins/Mesowx/meso/include/config-example.json',
                    'skins/Mesowx/meso/include/config.json.tmpl',
                    'skins/Mesowx/meso/include/config-RemoteSync.json.tmpl',
                    'skins/Mesowx/meso/include/Entity.class.php',
                    'skins/Mesowx/meso/include/EntityRetentionPolicy.class.php',
                    'skins/Mesowx/meso/include/EntityRetentionPolicyFactory.class.php',
                    'skins/Mesowx/meso/include/HttpUtil.class.php',
                    'skins/Mesowx/meso/include/JsonConfig.class.php',
                    'skins/Mesowx/meso/include/JsonUtil.class.php',
                    'skins/Mesowx/meso/include/PDOConnectionFactory.class.php',
                    'skins/Mesowx/meso/include/TableEntity.class.php',
                    'skins/Mesowx/meso/include/Unit.class.php',
                    'skins/Mesowx/meso/include/WindowRetentionPolicy.class.php',
                    'skins/Mesowx/meso/js/AbstractHighstockChart.js',
                    'skins/Mesowx/meso/js/AbstractRealTimeRawDataProvider.js',
                    'skins/Mesowx/meso/js/AggregateDataProvider.js',
                    'skins/Mesowx/meso/js/ChangeIndicatedValue.js',
                    'skins/Mesowx/meso/js/MesoConsole.js',
                    'skins/Mesowx/meso/js/meso.js',
                    'skins/Mesowx/meso/js/PollingRealTimeRawDataProvider.js',
                    'skins/Mesowx/meso/js/SocketIoRealTimeRawDataProvider.js',
                    'skins/Mesowx/meso/js/StatsDataProvider.js',
                    'skins/Mesowx/style/mesowx.css',
                    'skins/Mesowx/README.html',
                    'skins/Mesowx/footnote.inc',
                    'skins/Mesowx/links.inc',
                    'skins/Mesowx/index.html.tmpl'])]
            )
